File: case_studies/wcbgb/library/src/sensemaker_utils.ts
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Simple utils.

import { Comment, CommentRecord, SummaryContent, Topic } from "./types";
import { RETRY_DELAY_MS } from "./models/model_util";
import { voteInfoToString } from "./tasks/utils/citation_utils";

/**
 * Rerun a function multiple times.
 * @param func the function to attempt
 * @param isValid checks that the response from func is valid
 * @param maxRetries the maximum number of times to retry func
 * @param errorMsg the error message to throw
 * @param retryDelayMS how long to wait in miliseconds between calls
 * @param funcArgs the args for func and isValid
 * @param isValidArgs the args for isValid
 * @returns the valid response from func
 */
/* eslint-disable  @typescript-eslint/no-explicit-any */
export async function retryCall<T>(
  func: (...args: any[]) => Promise<T>,
  isValid: (response: T, ...args: any[]) => boolean,
  maxRetries: number,
  errorMsg: string,
  retryDelayMS: number = RETRY_DELAY_MS,
  funcArgs: any[],
  isValidArgs: any[]
) {
  /* eslint-enable  @typescript-eslint/no-explicit-any */
  for (let attempt = 1; attempt <= maxRetries; attempt++) {
    try {
      const response = await func(...funcArgs);
      if (isValid(response, ...isValidArgs)) {
        return response;
      }
      console.error(`Attempt ${attempt} failed. Invalid response:`, response);
      /* eslint-disable  @typescript-eslint/no-explicit-any */
    } catch (error: any) {
      if (
        error.message?.includes("Too Many Requests") ||
        error.code === 429 ||
        error.status === "RESOURCE_EXHAUSTED"
      ) {
        // log error message only to avoid spamming the logs with stacktraces
        console.error(`Attempt ${attempt} failed: ${error.message}`);
      } else {
        console.error(`Attempt ${attempt} failed:`, error);
      }
    }

    // Exponential backoff calculation
    const backoffGrowthRate = 1; // controls how quickly delay increases b/w retries (higher value = faster increase)
    const delay = retryDelayMS * Math.pow(backoffGrowthRate, attempt - 1);
    console.log(`Retrying in ${delay / 1000} seconds (attempt ${attempt})`);
    await new Promise((resolve) => setTimeout(resolve, delay));
  }
  throw new Error(`Failed after ${maxRetries} attempts: ${errorMsg}`);
}

/**
 * Combines the data and instructions into a prompt to send to Vertex.
 * @param instructions: what the model should do.
 * @param data: the data that the model should consider.
 * @param additionalContext additional context to include in the prompt.
 * @param dataWrapper: a function for wrapping each data entry
 * @returns the instructions and the data as a text
 */
export function getAbstractPrompt<T>(
  instructions: string,
  data: T[],
  dataWrapper: (data: T) => string,
  additionalContext?: string
) {
  return `
<instructions>
  ${instructions}
</instructions>
${additionalContext ? "\n<additionalContext>\n  " + additionalContext + "\n</additionalContext>\n" : ""}
<data>
  ${data.map(dataWrapper).join("\n  ")}
</data>`;
}

/**
 * Combines the data and instructions into a prompt to send to Vertex.
 * @param instructions: what the model should do.
 * @param data: the data that the model should consider.
 * @param additionalContext additional context to include in the prompt.
 * @returns the instructions and the data as a text
 */
export function getPrompt(
  instructions: string,
  data: string[],
  additionalContext?: string
): string {
  return getAbstractPrompt(
    instructions,
    data,
    (data: string) => `<comment>${data}</comment>`,
    additionalContext
  );
}

/**
 * Utility function for formatting the comments together with vote tally data
 * @param commentData: the data to summarize, as an array of Comment objects
 * @returns: comments, together with vote tally information as JSON
 */
export function formatCommentsWithVotes(commentData: Comment[]): string[] {
  return commentData.map(
    (comment: Comment) =>
      comment.text + "\n      vote info per group: " + JSON.stringify(comment.voteInfo)
  );
}

/**
 * Converts the given commentRecords to Comments.
 * @param commentRecords what to convert to Comments
 * @param missingTexts the original comments with IDs match the commentRecords
 * @returns a list of Comments with all possible fields from commentRecords.
 */
export function hydrateCommentRecord(
  commentRecords: CommentRecord[],
  missingTexts: Comment[]
): Comment[] {
  const inputCommentsLookup = new Map<string, Comment>(
    missingTexts.map((comment: Comment) => [comment.id, comment])
  );
  return commentRecords
    .map((commentRecord: CommentRecord): Comment | undefined => {
      // Combine the matching Comment with the topics from the CommentRecord.
      const comment = inputCommentsLookup.get(commentRecord.id);
      if (comment) {
        comment.topics = commentRecord.topics;
      }
      return comment;
    })
    .filter((comment: Comment | undefined): comment is Comment => {
      return comment !== undefined;
    });
}

/**
 * Groups categorized comments by topic and subtopic.
 *
 * @param categorized An array of categorized comments.
 * @returns A JSON representing the comments grouped by topic and subtopic.
 *
 * Example:
 * {
 *   "Topic 1": {
 *     "Subtopic 2": {
 *       "id 1": "comment 1",
 *       "id 2": "comment 2"
 *     }
 *   }
 * }
 *
 * TODO: create a similar function to group comments by topics only.
 */
export function groupCommentsBySubtopic(categorized: Comment[]): {
  [topicName: string]: {
    [subtopicName: string]: { [commentId: string]: Comment };
  };
} {
  const groupedComments: {
    [topicName: string]: {
      [subtopicName: string]: { [commentId: string]: Comment };
    };
  } = {};
  for (const comment of categorized) {
    if (!comment.topics || comment.topics.length === 0) {
      console.log(`Comment with ID ${comment.id} has no topics assigned.`);
      continue;
    }
    for (const topic of comment.topics) {
      if (!groupedComments[topic.name]) {
        groupedComments[topic.name] = {}; // init new topic name
      }
      if ("subtopics" in topic) {
        for (const subtopic of topic.subtopics || []) {
          if (!groupedComments[topic.name][subtopic.name]) {
            groupedComments[topic.name][subtopic.name] = {}; // init new subtopic name
          }
          groupedComments[topic.name][subtopic.name][comment.id] = comment;
        }
      }
    }
  }
  return groupedComments;
}

/**
 * Gets a set of unique topics and subtopics from a list of comments.
 * @param comments the comments with topics and subtopics to consider
 * @returns a set of unique topics and subtopics
 */
export function getUniqueTopics(comments: Comment[]): Topic[] {
  const topicNameToTopic = new Map<string, Topic>();
  for (const comment of comments) {
    if (comment.topics) {
      for (const topic of comment.topics) {
        const existingTopic = topicNameToTopic.get(topic.name);
        if (!existingTopic) {
          topicNameToTopic.set(topic.name, topic);
        } else {
          const existingSubtopics =
            "subtopics" in existingTopic
              ? existingTopic.subtopics.map((subtopic) => subtopic.name)
              : [];
          const newSubtopics =
            "subtopics" in topic ? topic.subtopics.map((subtopic) => subtopic.name) : [];
          const uniqueSubtopics = new Set([...existingSubtopics, ...newSubtopics]);
          topicNameToTopic.set(topic.name, {
            name: topic.name,
            subtopics: Array.from(uniqueSubtopics).map((subtopic) => ({ name: subtopic })),
          });
        }
      }
    }
  }
  return Array.from(topicNameToTopic.values());
}

/**
 * Format a decimal number as a percent string with the given precision
 * @param decimal The decimal number to convert
 * @param precision The precision
 * @returns A string representing the equivalent percentage
 */
export function decimalToPercent(decimal: number, precision: number = 0): string {
  const percentage = decimal * 100;
  const roundedPercentage = Math.round(percentage * 10 ** precision) / 10 ** precision;
  return `${roundedPercentage}%`;
}

/**
 * Interface for specifying an extra column for a markdown table, as a columnName and
 * getValue function.
 */
export interface ColumnDefinition {
  columnName: string;
  getValue: (comment: Comment) => any;
}

/**
 * Return the markdown corresponding to the extraColumns specification for the given Comment row,
 * without either the leading or tailing | bars.
 * @param extraColumns Either a Comment object key (string) or ColumnDefinition object
 * @param row Comment object
 * @returns A string representing the additional column values
 */
function extraColumnDataMd(
  extraColumns: (keyof Comment | ColumnDefinition)[],
  row: Comment
): string {
  return extraColumns.length > 0
    ? " <small>" +
        extraColumns
          .map((extraColumn) => columnValue(extraColumn, row))
          .join("</small> | <small>") +
        "</small> |"
    : "";
}

/**
 * Returns the table cell entry for the given ColumnDefinition (or Comment key) and Comment
 * object.
 * @param extraColumn Either a Comment key, or a ColumnDefinition object
 * @param comment A comment object
 * @returns The corresponding table cell value
 */
function columnValue(extraColumn: keyof Comment | ColumnDefinition, comment: Comment) {
  return typeof extraColumn === "string" ? comment[extraColumn] : extraColumn.getValue(comment);
}

/**
 * Return header name for extraColumn specification, either the returning the string back
 * or getting the columnName specification for a ColumnDefinition object.
 */
function columnHeader(extraColumn: string | ColumnDefinition) {
  return typeof extraColumn === "string" ? extraColumn : extraColumn.columnName;
}

/**
 * Returns a markdown table of comment data for inspection and debugging.
 * @param comments An array of Comment objects to include in the table.
 * @param extraColumns An array of keys of the comment objects to add as table cells.
 * @returns A string containing the markdown table.
 */
export function commentTableMarkdown(
  comments: Comment[],
  extraColumns: (keyof Comment | ColumnDefinition)[] = []
): string {
  // Format the comments as a markdown table, with rows keyed by comment id,
  // displaying comment text and vote tally breakdown.
  const hasExtraCols = extraColumns.length > 0;
  const extraHeaders = extraColumns.map(columnHeader);
  const extraHeadersMd = hasExtraCols ? " " + extraHeaders.join(" | ") + " |" : "";
  const extraHeadersUnderlineMd = hasExtraCols
    ? " " + extraHeaders.map((h) => "-".repeat(h.length)).join(" | ") + " |"
    : "";
  return (
    `\n| id | text | votes |${extraHeadersMd}\n| -- | ---- | ---- |${extraHeadersUnderlineMd}\n` +
    comments.reduce(
      (ct: string, comment: Comment): string =>
        ct +
        `| ${comment.id}&nbsp; | ${comment.text} | <small>${voteInfoToString(comment)}</small> |${extraColumnDataMd(extraColumns, comment)}\n`,
      ""
    )
  );
}

/**
 * Executes a batch of asynchronous functions (callbacks) concurrently.
 * This is essential for running multiple LLM calls in parallel, as it submits requests downstream as a batch.
 *
 * @param callbacks A batch of functions, each of which returns a Promise<T>.
 * @returns A Promise that resolves to an array containing the resolved values of the
 * promises returned by the callbacks, in the same order as the callbacks.
 */
export async function executeConcurrently<T>(callbacks: (() => Promise<T>)[]): Promise<T[]> {
  // NOTE: if a least one callback fails, the entire batch fails.
  // Because of that, we should aim to retry any failed callbacks down the call stack,
  // and avoid retries higher up the stack, as it will retry entire batch from scratch, including completed callbacks.
  return await Promise.all(callbacks.map((callback) => callback()));
}

/**
 * This function creates a copy of the input summaryContent object, filtering out
 * any subContents according to filterFn, as appropriate
 * @param summaryContent Input summary content
 * @returns the resulting summary conten, as a new data structure
 */
export function filterSummaryContent(
  summaryContent: SummaryContent,
  filterFn: (s: SummaryContent) => boolean
): SummaryContent {
  const filteredTopicSummary: SummaryContent = {
    title: summaryContent.title,
    text: summaryContent.text,
    citations: summaryContent.citations,
    subContents: summaryContent.subContents
      ?.filter(filterFn)
      .map((s: SummaryContent) => filterSummaryContent(s, filterFn)),
  };
  return filteredTopicSummary;
}

File: case_studies/wtp/sensemaker_utils.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Common utils for Sensemaker."""

import asyncio
import html
import logging
import random
import re
from typing import List, Callable, TypeVar, Any, Coroutine, Optional

RETRY_DELAY_SEC = 10


_ALLOWED_TAG_REGEX = re.compile(r"(</?(?:question|response|quote)>)")


def get_prompt(
    instructions: str,
    data: List[str],
    additional_context: Optional[str] = None,
) -> str:
  """
  Combines the data and instructions into a prompt to send to an LLM.
  Each data item is wrapped in <statement> tags.

  Args:
      instructions: What the model should do.
      data: The data that the model should consider (list of strings, typically statements).
      additional_context: Additional context to include in the prompt.
  Returns:
      The instructions and the data as a text.
  """
  cleaned_data = []
  for item in data:
    # We need to escape '<' and '>' characters that are not part of an allowed tag.
    # We do this by splitting the string by the allowed tags, and then escaping the parts in between.
    parts = _ALLOWED_TAG_REGEX.split(item)
    escaped_parts = []
    for i, part in enumerate(parts):
      if i % 2 == 1:  # This is a tag, so we leave it as is.
        escaped_parts.append(part)
      else:  # This is not a tag, so we escape it.
        escaped_parts.append(html.escape(part))
    cleaned_data.append("".join(escaped_parts))

  wrapped_data = [f"<item>{item}</item>" for item in cleaned_data]

  return f"""
{f'''<additionalContext>
  {additional_context}
</additionalContext>''' if additional_context else ""}

<instructions>
  {instructions}
</instructions>

<data>
{'''
'''.join(wrapped_data)}
</data>"""


# A generic type variable used to ensure that the type of an argument is the
# same as the type of a return value in functions like `retry_call` and
# `execute_concurrently`.
T = TypeVar("T")


async def execute_concurrently(
    callbacks: List[Callable[[], Coroutine[Any, Any, T]]],
) -> List[T]:
  """
  Executes a list of awaitable tasks (callbacks that return coroutines) concurrently.

  Args:
      callbacks: A list of functions, where each function returns a coroutine.
                 Example: [functools.partial(async_func1), functools.partial(async_func2)]
  Returns:
      A list containing the resolved values of the coroutines,
      in the same order as the input callbacks.
  """
  # Create coroutine objects by calling the functions in the callbacks list
  coroutines_to_run: List[Coroutine[Any, Any, T]] = [cb() for cb in callbacks]

  # NOTE: If at least one callback fails, the entire batch fails.
  # Retries should happen within the individual callbacks if needed.
  results: List[T] = await asyncio.gather(
      *coroutines_to_run, return_exceptions=False
  )
  return results


async def retry_call(
    func: Callable[..., Coroutine[Any, Any, T]],
    is_valid: Callable[[T, ...], bool],
    max_retries: int,
    error_msg: str,
    retry_delay_sec: int = RETRY_DELAY_SEC,
    func_args: Optional[List[Any]] = None,
    is_valid_args: Optional[List[Any]] = None,
) -> T:
  """
  Reruns an async function multiple times if validation fails or an exception occurs.

  Args:
      func: The async function to attempt.
      is_valid: A function that checks if the response from func is valid.
      max_retries: The maximum number of times to retry func.
      error_msg: The error message to raise if all retries fail.
      retry_delay_sec: How long to wait in seconds between calls.
      func_args: Arguments for func.
      is_valid_args: Arguments for is_valid (after the response argument).
  Returns:
      The valid response from func.
  Raises:
      Exception if all retries fail.
  """
  func_args = func_args or []
  is_valid_args = is_valid_args or []
  current_delay_sec = float(retry_delay_sec)

  for attempt in range(1, max_retries + 1):
    try:
      response = await func(*func_args)
      if is_valid(response, *is_valid_args):
        return response
      logging.warning(
          f"Attempt {attempt} failed validation. Invalid response: {response}"
      )
    except Exception as e:
      error_str = str(e).lower()
      if (
          "too many requests" in error_str
          or "429" in error_str  # HTTP code for too many requests
          or "resource_exhausted" in error_str
          or "service is currently unavailable" in error_str
          or "internal error" in error_str
          or "deadline exceeded" in error_str
      ):
        logging.warning(
            f"Attempt {attempt} failed with potentially transient error: {e}"
        )
      else:
        logging.error(
            f"Attempt {attempt} failed with error: {e}", exc_info=True
        )

    if attempt >= max_retries:
      break

    logging.info(
        f"Retrying in {current_delay_sec} seconds (attempt"
        f" {attempt} of {max_retries})..."
    )
    await asyncio.sleep(current_delay_sec)
    # Exponential backoff with jitter
    current_delay_sec = min(current_delay_sec * 1.5, 60.0)  # Cap at 60 seconds
    current_delay_sec += random.uniform(0, 0.1 * current_delay_sec)

  raise Exception(f"Failed after {max_retries} attempts: {error_msg}")
